### PR TITLE
adding profile interpolation

### PR DESCRIPTION
**Pull request by elizabethteng**
see posydon/interpolation/profile_interpolation.py. Includes funcitonality for Density and H mass fraction profiles.

### CODE REVIEW

File: posydon/interpolation/profile_interpolation.py [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+        """Extracts profile data from '.h5' grid files and saves to file.
+        Args:
+            train_path (str) : path/name of '.h5' file for training data.
+            valid_path (str) : path/name of '.h5' file for testing/validation data.
+            profile_names (array-like) : list of profile quantities to be extracted.
+        """
+        self.names = profile_names
+        
+        try:
+            print(valid.final_values["S1_state"][0])
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: Did I missed it, or is the variable `valid` not defined before that try at all? It will get set in line 45.

Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py]: yes, you're right! thanks for catching that

File: posydon/interpolation/profile_interpolation.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+        Returns:
+            scalars (array-like) : dictionary containing initial m1, m2, p and final s1_state, final_mass.
+            profiles (array-like) : all N specified profiles, shape (N,200).
+        """
+        # open individual run as a DataFrame
+        df = pd.DataFrame(grid[ind]['final_profile1'])
+        df = df.sort_values(by="mass")
+        df = df.reset_index(drop=True)
+
+        # grab input values, final star 1 state, final star 1 mass
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py:95]: Minor: If you need the values only for putting them in the dictionary, there is no need to define other variables first (this only wastes resources).

File: posydon/interpolation/profile_interpolation.py [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+        scalars = {"m1":m1,
+                   "m2":m2,
+                   "p":p,
+                   "s1_state":state,
+                   "final_mass":final_mass}
+
+        # grab output vectors, interpolate to normalize
+        profiles=np.zeros([len(self.names),200])
+        for i,prof in enumerate(self.names):
+            f = interp1d(df['mass']/final_mass,df[prof],fill_value="extrapolate")
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: You may want to check first, that the `df` contains the column `prof` and add a catch if it is not there.

File: posydon/interpolation/profile_interpolation.py [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+        self.profiles = []
+        training_failed = []
+        
+        for i in range(len(train)):
+            try:
+                scalars,profiles = self.scrape(train,i)
+                self.scalars = self.scalars.append(scalars,ignore_index=True)
+                self.profiles.append(profiles)
+            except:
+                training_failed.append(i)
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py:75]: `training_failed` and similarly `testing_failed` is created, but never used. I guess, you planed to raise a warning at the end of the initialization, telling which runs in the psygrid failed. Please, add this warning output.

File: posydon/interpolation/profile_interpolation.py [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+            self.valid_profiles (array-like) : M profiles for each of N testing binaries, shape (N,M,200).
+            self.valid_scalars (array-like) : DataFrame containing initial and final conditions for N testing binaries.
+        """
+        with open(filename, 'rb') as f:
+            myattrs = pd.read_pickle(f)
+            for key in myattrs:
+                setattr(self, key, myattrs[key])
+        self.profiles = np.array(self.profiles)
+        self.valid_profiles = np.array(self.valid_profiles)
+        return self.profiles,self.scalars,self.valid_profiles,self.valid_scalars
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: I'm not sure, whether it is a good idea to return here this 4 parts. Wouldn't is be better to return `self`, in this way you wouldn't discard other key (you may add in the future). And you won't need to create new objects on the fly.
But I see, where you called it you don't use the return value at all.

File: posydon/interpolation/profile_interpolation.py [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+        m1 = grid.initial_values["star_1_mass"][ind]
+        m2 = grid.initial_values["star_2_mass"][ind]
+        p = grid.initial_values["period_days"][ind]
+        state = grid.final_values["S1_state"][ind]  
+        final_mass = grid.final_values["star_1_mass"][ind]
+        scalars = {"m1":m1,
+                   "m2":m2,
+                   "p":p,
+                   "s1_state":state,
+                   "final_mass":final_mass}
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: Should we call `final_mass` better `reference_mass` or `unit_mass`? The term `final_mass` won't tell which star's final mass it is, while we do have two final star masses.

**Comment by mkruckow** [comment]
I talked to Simone, he said, that we need the angular momentum profile to be well for calculating the spins. Could you please think about adding a feature to get a new profile from other profiles.
The problem is the following: Let's define a function `f(a,b)`, which calculates a quantity based on `a` and `b`. Furthermore, we have the interpolation mapping `i(x)` which interpolates one quantity. Now f(i(a),i(b))!=i(f(a,b)), while it should be equal if we would have detailed calculations instead of interpolation. Because of the inequality, we need to think about, which quantities should be interpolated (usually, the ones we need the best accuracy on).
Another example of this is the interpolation of X, Y and Z, because 1=X+Y+Z should always hold.

File: posydon/interpolation/profile_interpolation.py [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+            profiles (array-like) : N profiles to be post-processed for monotonicity
+        Returns:
+            profiles_mono (array-like) : post-processed profiles
+        """
+        
+        def mono_renorm(arr):
+            arr_copy = arr.copy()
+            # force rising points down
+            for i in range(1,len(arr_copy)):
+                if arr_copy[i]-arr_copy[i-1]>0:
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: Minor: Why don't you write it as `if arr_copy[i]>arr_copy[i-1]:`? There shouldn't be a need to request an additional difference to be calculated.

File: posydon/interpolation/profile_interpolation.py [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+                    arr_copy[i]=arr_copy[i-1]
+            # cut off points below surface value
+            low = np.where(arr_copy<arr[-1])[0]
+            arr_copy[low] = arr[-1]
+            return arr_copy
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: Minor: Couldn't lines 237-239 be better written as `return np.where(arr_copy<arr[-1], arr[-1], arr_copy)`?

**Comment by mkruckow** [comment]
In general, whenever you use an imported function, could you please write a comment there, what this function is supposed to do (and hopefully is doing in all cases).

File: posydon/interpolation/profile_interpolation.py [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+                             'H-rich_Central_He_depleted',
+                             'H-rich_non_burning',
+                             'H-rich_Core_H_burning',
+                             'H-rich_Core_He_burning',
+                             'stripped_He_Central_He_depleted',
+                             'stripped_He_non_burning',
+                             'NS',
+                             'stripped_He_Central_C_depletion',
+                             'H-rich_Central_C_depletion',
+                             'BH']
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: How do you selected those states? E.g. why is there no "WD"?

File: posydon/interpolation/profile_interpolation.py [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+        Returns:
+            b_models (array-like) : dictionary containing boundary models.
+        """
+        b_models = {}
+        for state in ["H-rich_Shell_H_burning",
+                  'H-rich_Core_H_burning',
+                  'H-rich_Core_He_burning', 
+                  "H-rich_Central_He_depleted",
+                  'H-rich_Central_C_depletion',
+                  "NS","BH"]:
```
Comment (mkruckow) [inline posydon/interpolation/profile_interpolation.py]: Minor: You shouldn't mix different string types (single vs. double quote).

**Comment by ssbvr** [comment]
@elizabethteng I would like to check the quality of a sample of interpolated profiles in the black hole spin calculation context. Could you please find a couple of profiles in the grid you are interpolating where `grid.final_values["Patton&Sukhbold20-engineN20_state"] == 'BH'` and `grid.final_values["Patton&Sukhbold20-engineN20_state_spin"]` is large, say > 0.5 (if you are interpolating CO-HeMS grids) or 0.25 (if you are interpolating HMS-HMS grids)? I will later test the spin prediction accuracy from the `posydon/binary_evol/SN/profile_collapse.py` calculation on the ORIGINAL, LITE, and interpolated profiles.

File: posydon/interpolation/profile_interpolation.py (hunk 9 above)
Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py]: I just used the ones that showed up in the grid I used to develop this; do you know if others are missing besides "WD"?

File: posydon/interpolation/profile_interpolation.py (hunk 8 above)
Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py]: I didn't know you could do that! will definitely be very helpful

File: posydon/interpolation/profile_interpolation.py (hunk 9 above)
Comment (mkruckow) [reply posydon/interpolation/profile_interpolation.py]: from plot_defaults.py:
`'H-rich_Core_C_burning'`
`'stripped_He_Core_He_burning'`
`'stripped_He_Core_C_burning'`
There are some more strange states, which should never happen:
`'stripped_He_Core_H_burning'`
`'stripped_He_Shell_H_burning'`
`'undetermined_evolutionary_state'`
By the way, I'd suggest you to bring them in a more logical order.

**Review by mkruckow — DISMISSED** [review]
Now, the code looks fine to me.
I didn't run it, hence I suggest some more test runs if not done already.

File: docs/interpolation/Profile-Interpolator/ProfileInterpolator.rst [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+===============================
+
+To load a pretrained interpolator we need to pass the optional
+``load_interpolator`` argument to the ``ProfileInterpolator`` 
+instance which specifies the path to a .pkl file where the 
+pretrained interpolator can be loaded from. 
+
+.. code-block:: python
+
+  model = ProfileInterpolator(load_interpolator = "/path/to/profile/interpolator.pkl")
```
Comment (philipp-rajah) [inline docs/interpolation/Profile-Interpolator/ProfileInterpolator.rst:51]: we should be consistent with how we load pertained interpolators across all of our interpolators. In the IFInterpolator we have a method called load that we call after constructing the class and the argument that points to where the interpolator is saved is called filename.

File: posydon/interpolation/profile_interpolation.py [hunk 12]
```diff
@@ -0,0 +1,572 @@
...
+# POSYDON
+from posydon.grids.psygrid import PSyGrid
+from posydon.interpolation.IF_interpolation import IFInterpolator
+
+# Math and ML
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+tf.get_logger().setLevel('ERROR')
+from tensorflow.keras import layers, losses, models, optimizers
```
Comment (philipp-rajah) [inline posydon/interpolation/profile_interpolation.py:20]: In the future we need to decide on either TensorFlow or PyTorch

File: posydon/interpolation/profile_interpolation.py [hunk 13]
```diff
@@ -0,0 +1,572 @@
...
+        Args:
+            grid (obj) : PSyGrid object.
+            ind (int) : index of run to be scraped.
+        Returns:
+            scalars (array-like) : dictionary containing initial 
+                                   m1, m2, p and final s1_state, final_m1.
+            profiles (array-like) : all N specified profiles, shape (N,200).
+        """
+        # open individual run as a DataFrame
+        df = pd.DataFrame(grid[ind]['final_profile1'])
```
Comment (philipp-rajah) [inline posydon/interpolation/profile_interpolation.py:91]: So I like the idea of using pandas df, but in IF mostly numpy structured arrays are used. I don't think this matters, but it may be worth discussing in the future, and deciding on one.

File: posydon/interpolation/profile_interpolation.py [hunk 14]
```diff
@@ -0,0 +1,572 @@
...
+        DONT_SAVE = []
+        myattrs = {key: SAVE_ATTRS[key]
+                   for key in SAVE_ATTRS if key not in DONT_SAVE}
+        with open(filename, 'wb') as f:
+            pickle.dump(myattrs, f)
+            
+            
+class ProfileInterpolator:
+    
+    def __init__(self,load_interpolator=None):
```
Comment (philipp-rajah) [inline posydon/interpolation/profile_interpolation.py]: name this argument filename to keep it consistent with the IFInterpolator and have a separate load method

File: posydon/interpolation/profile_interpolation.py [hunk 15]
```diff
@@ -0,0 +1,572 @@
...
+            loss (object) : loss function for training.
+        """
+        print("training on PCA weights...")
+        
+        self.model_prof.compile(optimizers.Adam(clipnorm=1),loss=loss)
+        callback = tf.keras.callbacks.EarlyStopping(monitor="loss",patience=200)
+        history = self.model_prof.fit(self.initial,self.weights,
+                                      epochs=3000,callbacks=[callback],verbose=0,
+                                      validation_data=(self.valid_initial,
+                                                       self.valid_weights))
```
Comment (philipp-rajah) [inline posydon/interpolation/profile_interpolation.py]: maybe name this PCA weights, pretty minor, but I was initially confused about the possibility of passing network weights since I don't work with tf. Also it would be useful to be able to see the model train and to see the loss history. Might make astro people less reluctant to use it since it's not just this black box. Look into programs like tensorboard.

File: posydon/interpolation/profile_interpolation.py [hunk 16]
```diff
@@ -0,0 +1,572 @@
...
+        """Trains NN models. 
+        Args: 
+            loss (object) : loss function for training.
+        """
+        print("training on PCA weights...")
+        
+        self.model_prof.compile(optimizers.Adam(clipnorm=1),loss=loss)
+        callback = tf.keras.callbacks.EarlyStopping(monitor="loss",patience=200)
+        history = self.model_prof.fit(self.initial,self.weights,
+                                      epochs=3000,callbacks=[callback],verbose=0,
```
Comment (philipp-rajah) [inline posydon/interpolation/profile_interpolation.py]: I would have more of these parameters be accessible for tweaking by the user. Depending on the data, it may take more than 3000 epochs to converge, and a different user may want to be more patient. I think that since this whole DL thing is new to most astro people, there will definitely be those curious in what's going on under the hood and some may want to play around with things, which is something I think we should encourage. This goes for all calls of the fit function.

File: posydon/interpolation/profile_interpolation.py [hunk 17]
```diff
@@ -0,0 +1,572 @@
...
+            # identify input/output testing data for class
+            valid_indices = self.valid_sort_ind[state]
+            valid_inputs = self.valid_initial[valid_indices]
+            valid_prof = self.valid_profiles[valid_indices]
+            
+            # calculate boundary points for training and testing data 
+            if "burning" in state:  # these classes' profile shapes have 2 boundary points
+                outs = 2
+                bounds = []
+                nonflat = []  # TODO: explain this
```
Comment (philipp-rajah) [inline posydon/interpolation/profile_interpolation.py]: what is nonflat?

File: docs/interpolation/Profile-Interpolator/ProfileInterpolator.rst [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+where the interpolator will be saved to.
+
+.. code-block:: python
+
+ model.save("path/for/profile/interpolator.pkl")
+   
+Evaluating on Testing Data
+==========================
+
+To evaluate the interpolator on the testing grid that was used as validation data in
```
Comment (philipp-rajah) [inline docs/interpolation/Profile-Interpolator/ProfileInterpolator.rst:94]: Validation set should not be the same as the testing set, this is a bigger issue I'd say. It's probably ok in this case because we won't ever see data far outside of the distribution of the testing and training sets. I'll double check with people here though.

**Review by philipp-rajah — COMMENTED** [review]
Made a lot of minor comments. I'm assuming that most of the logic of the code works and that this had been tested. Need to check with Aggelos about the validation and testing set.

File: posydon/interpolation/profile_interpolation.py (hunk 13 above)
Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py:91]: I'll make a note to bring this and tf/torch up in a dev meeting

File: posydon/interpolation/profile_interpolation.py (hunk 17 above)
Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py]: just added this comment: # ensures that training data only has the correct "non-flat" shape
\# to avoid issues with calculating the boundary points

File: docs/interpolation/Profile-Interpolator/ProfileInterpolator.rst (hunk 18 above)
Comment (elizabethteng) [reply docs/interpolation/Profile-Interpolator/ProfileInterpolator.rst:94]: yeah we should bring this up when I present on Monday

File: posydon/interpolation/profile_interpolation.py (hunk 16 above)
Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py]: I made the parameters for the PCA weights and H mass fraction boundary points customizable. I don't think anyone could ever need more for the minimum density, since it's such a simple problem. usually only takes a few seconds to train.

File: posydon/interpolation/profile_interpolation.py (hunk 15 above)
Comment (elizabethteng) [reply posydon/interpolation/profile_interpolation.py]: adding this to end of next PR task list

**Comment by elizabethteng** [comment]
did tests on these changes, everything works as expected

**Review by philipp-rajah — COMMENTED** [review]
I meant to mention this yesterday but forgot, what about GPU acceleration? I. don't really see it in the code

**Review by mkruckow — APPROVED** [review]

**Comment by elizabethteng** [comment]
> I meant to mention this yesterday but forgot, what about GPU acceleration? I. don't really see it in the code

I didn't use GPU acceleration because the training doesn't take that long. like 1 hour per grid maximum. tbh don't think it's worth the effort

**Review by philipp-rajah — APPROVED** [review]

**Review by ssbvr — APPROVED** [review]
